cpp/ql/src/experimental/Security/CWE/CWE-758/DangerousUseMallocWithZeroSize.ql

```diff
@@ -39,15 +40,21 @@ predicate existsChecksorSet(FunctionCall fc) {
       ae.getASuccessor*() = fc
     )
     or
-    exists(IfStmt ifs, ComparisonOperation co, Expr ec |
-      ifs.getCondition().getAChild*() = co and
-      co.hasOperands(fc.getArgument(0).(VariableAccess).getTarget().getAnAccess(), ec) and
-      exists(Expr etmp |
-        etmp.getValue() = ["0", "1"] and
+    exists(GuardCondition gc, Expr bound, Expr val |
+      (
+        globalValueNumber(fc.getArgument(0)) = globalValueNumber(bound) or
+        fc.getArgument(0).(VariableAccess).getTarget() = bound.(VariableAccess).getTarget()
+      ) and
+      (
+        val.getValue() = ["0", "1"] and
         (
-          ec = etmp or
-          globalValueNumber(ec) = globalValueNumber(etmp)
+          gc.ensuresEq(bound, val, _, fc.getBasicBlock(), _) or
+          gc.ensuresEq(val, bound, _, fc.getBasicBlock(), _) or
+          gc.ensuresLt(bound, val, _, fc.getBasicBlock(), _) or
+          gc.ensuresLt(val, bound, _, fc.getBasicBlock(), _)
         )
```

Comment:
Watch out here. You've declared `val` further up, but you only bind the variable in this disjunction. Perhaps you want to replace this snippet with:
```ql
exists(Expr val | val.getValue() = ["0", "1"] |
  gc.ensuresEq(bound, val, _, fc.getBasicBlock(), _) or
  gc.ensuresEq(val, bound, _, fc.getBasicBlock(), _) or
  gc.ensuresLt(bound, val, _, fc.getBasicBlock(), _) or
  gc.ensuresLt(val, bound, _, fc.getBasicBlock(), _)
)
```

Reply:
Thank you.
I have corrected.
beautiful idea.